### ISSUE TITLE
Infinite loop for some recurrence rule

### ISSUE THREAD

**Issue by asvahagnas** [node-schedule/node-schedule#273]
If set recurrence rule this values {"year":2016,"month":3,"date":29,"hour":12,"minute":2,"second":0}, node freezing and cpu reaching top loads.

**Comment by santigimeno** [comment]
Confirmed. I'll try to look into it. Thanks for reporting this

**Comment by Ampakinetic** [comment]
I've just started getting this too, very odd

**Comment by Ampakinetic** [comment]
I've traced the loop with the following test:

```
testSchedule: function () {
		var ctx = {
			year: 2016,
			month: 3,
			day: 30,
			hour: 9,
			minute: 5,
			second: 0
		};

		var job = new schedule.Job(ctx, function () {
			schedule.cancelJob(job);
			console.log("ran job");
			
		});
		
		job.ctx = ctx;
		console.log("pre schedule");
		job.schedule(ctx);
		console.log("post schedule");
		
	}
```
The loop is in 
RecurrenceRule.prototype.nextInvocationDate on line 335

and the reason it's happening is because next.addMonth(); gets called repeatedly to get this.month and next.getMonth() to match.  Going round December this way appears to increment the year, which then starts the infinite loop as the year gets incremented forever because it no longer matches this.year.

Still looking into what the best fix is but posting this in case it will help get it sorted sooner.

**Comment by Ampakinetic** [comment]
I think I've fixed it with the following change to stop the year loop.  I'm not sure if the other time intervals have the same incrementing behaviour, so monthIsSet etc may be needed as well?

```
var yearIsSet = false;
	
	while (true) {
		if (this.year != null && !recurMatch(next.getFullYear(), this.year) && !yearIsSet) {
			next.addYear();
			next.setMonth(0);
			next.setDate(1);
			next.setHours(0);
			next.setMinutes(0);
			next.setSeconds(0);
			continue;
		} else {
			yearIsSet = true;
		}
                ....
```

**Comment by santigimeno** [comment]
@Ampakinetic thanks for the pointer. It has been helpful!

@asvahagnas can you check if https://github.com/node-schedule/node-schedule/pull/274 fixes the problem for you?. I'll update the PR in the following days with tests

**Comment by asvahagnas** [comment]
@santigimeno yes, thanks a lot.

**Comment by santigimeno** [comment]
Fixed in https://github.com/node-schedule/node-schedule/commit/0883f24dcb2717f04aba9bb8503f260c706ae144 and published as `1.1.1`
